src/loss.py

```diff
@@ -1,2 +1,32 @@
 from abc import ABC
+import numpy as np
 
+
+class Purity_function(ABC):
```

Comment:
class names should use Pascal case. Remember to refactor this in all of your code (same for other comments below)
```suggestion
class PurityFunction(ABC):
```